### PR TITLE
Scheduler separation from integration logic

### PR DESCRIPTION
**Pull request by jjsjann123**
Separate scheduling from integration.
Uses ExpressionEvaluator to infer run-time launch configuration

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h]: This results in two lookups, which can be avoided by using find:
```
const auto it = launch_configs_.find(type);
return it != launch_configs_.end() ? it->second : nullptr;
```

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h:305]: 1. Can we have the launch config outside, and independent of the Fusion? 
2. map to `const Val*`? (or even better, directly to `Int::ScalarType`)

File: torch/csrc/jit/codegen/cuda/kernel.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: the printfs are temporary, right?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:15]: please use static functions instead of anonymous namespaces

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: these could be const

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: = nullptr ?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: if we always map integers why not have the launch config map to a Int::ScalarType instead of a Int* ?

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/fusion.h:305]: Regarding 1. This is the part I'm still thinking about. I wouldn't mind having this outside of the `Fusion` with all arith nodes. (a separate `Fusion` with only the inputs & config logic nodes); But I don't know if we want to have a map of pointer to nodes, whose lifetime is owned by a separate `fusion` object. I could have them bounded together by a wrapper object tho.
Curious on the motivation of the separation? Are we worried about overhead?

2. `const` would make sense, but ExpressionEvaluator::evaluate takes `Val*` so we have to const_cast the specifier away, maybe we can update the function signature? https://github.com/csarofeen/pytorch/blob/481afe10e5d967f7800cde69cde70c82dde50c46/torch/csrc/jit/codegen/cuda/expr_evaluator.h#L50-L52

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Are we suggesting to have the `ExpressionEvaluator::evaluate` wrapped inside this as well? We still need to pass a `Int*`/`Val*` through the API, as later `ExpressionEvaluator::evaluator(Val* val, const EvaluationContext* context)` takes argument `Val*`.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp:15]: never mind, based on the conversation today, this is not the style guideline

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp:15]: Per recent discussion, anonymous namespaces are the default now.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Discussed offline. launch config should be inferred at run-time, so we need pointer to nodes.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:51]: The implementation here is not relevant. @kevinstephano 's new scheduler should be swapped in here at some point.
This is going to be removed soon-ish, so I don't see much value in refactoring this to separate the module for PW/fcd_reduction/non_fcd_reduction.

File: torch/csrc/jit/codegen/cuda/fusion.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:82]: nit: empty line after this swap

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h]: TORCH_CHECK(val->fusion() == this); ?

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h]: const auto?

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/fusion.h]: const auto?

File: torch/csrc/jit/codegen/cuda/kernel.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: nit: 
1. please empty line before the lambda definition
2. `-> int` is not required

File: torch/csrc/jit/codegen/cuda/kernel.cpp
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: Nit: I'd use TORCH_CHECK() - this is a contract with an external component, not an internal invariant

**Review by tlemo — CHANGES_REQUESTED** [review]
Looks good to me

Just one thing: please update the FusionClear/Copy/Move test cases to do a minimal checking that the launch_config_ is handled properly

**Comment by jjsjann123** [comment]
Note to myself. DO NOT MERGE TO DEV BRANCH until https://github.com/pytorch/pytorch/pull/40864 goes in.

**Comment by jjsjann123** [comment]
@tlemo I modified the tests slightly to check for `launch_configs_`
I'm not going to merge this because of on-going PR. But I would like to still get an approval stamp from you. so later I can merge this in later :)

**Review by tlemo — APPROVED** [review]
Thanks for updating the tests. LGTM!

**Review by kevinstephano — APPROVED** [review]
Looks good to me.

**Comment by jjsjann123** [comment]
Rebased to the new branch, passing local tests. I'm merging this guy since there's no changes from the approved PR.

**Pull request merged**
